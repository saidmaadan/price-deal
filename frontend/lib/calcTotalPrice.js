export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartProduct) => {
    if (!cartProduct.product) return tally;
    return tally + cartProduct.quantity * cartProduct.product.price;
  }, 0);
}
