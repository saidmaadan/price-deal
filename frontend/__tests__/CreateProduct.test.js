import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateProduct, { CREATE_PRODUCT_MUTATION } from '../components/CreateProduct';
import { fakeProduct } from '../lib/testUtils';

const storeCatelog = 'https://storecatelog.com/storecatelog.jpg';

// mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: storeCatelog,
    eager: [{ secure_url: storeCatelog }],
  }),
});

describe('<CreateProduct/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
          <MockedProvider>
            <CreateProduct />
          </MockedProvider>
        );
        const form = wrapper.find('form[data-test="form"]');
        expect(toJSON(form)).toMatchSnapshot();
    });

    it('uploads a file when changed', async () => {
        const wrapper = mount(
          <MockedProvider>
            <CreateProduct />
          </MockedProvider>
        );
        const input = wrapper.find('input[type="file"]');
        input.simulate('change', { target: { files: ['fakedog.jpg'] } });
        await wait();
        const component = wrapper.find('CreateProduct').instance();
        expect(component.state.image).toEqual(storeCatelog);
        expect(component.state.largeImage).toEqual(storeCatelog);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it('handles state updating', async () => {
        const wrapper = mount(
          <MockedProvider>
            <CreateProduct />
          </MockedProvider>
        );
        wrapper.find('#title').simulate('change', { target: { value: 'Testing', name: 'title' } });
        wrapper
          .find('#price')
            .simulate('change', { target: { value: 50000, name: 'price', type: 'number' } });
            wrapper
            .find('#originalPrice')
                .simulate('change', { target: { value: 80000, name: 'originalPrice', type: 'number' } });
                wrapper
                .find('#discount')
                .simulate('change', { target: { value: 20, name: 'discount', type: 'number' } });
        wrapper
          .find('#description')
          .simulate('change', { target: { value: 'This is a really nice product', name: 'description' } });
    
        expect(wrapper.find('CreateProduct').instance().state).toMatchObject({
            title: 'Testing',
            price: 50000,
            originalPrice: 80000,
            discount: 20,
            description: 'This is a really nice product',
        });
    });

    it('creates a product when the form is submitted', async () => {
        const product = fakeProduct();
        const mocks = [
          {
            request: {
              query: CREATE_PRODUCT_MUTATION,
              variables: {
                title: product.title,
                description: product.description,
                image: '',
                largeImage: '',
                price: product.price,
                originalPrice: product.originalPrice,
                discount: product.discount,
                
              },
            },
            result: {
              data: {
                createProduct: {
                  ...fakeProduct,
                  id: 'abc123',
                  __typename: 'Product',
                },
              },
            },
          },
        ];
    
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
              <CreateProduct />
            </MockedProvider>
          );
          // simulate someone filling out the form
        wrapper.find('#title').simulate('change', { target: { value: product.title, name: 'title' } });
        wrapper
            .find('#price')
            .simulate('change', { target: { value: product.price, name: 'price', type: 'number' } });
        wrapper
            .find('#originalPrice')
            .simulate('change', { target: { value: product.originalPrice, name: 'originalPrice', type: 'number' } });
        wrapper
            .find('#discount')
            .simulate('change', { target: { value: product.discount, name: 'discount', type: 'number' } });
        wrapper
            .find('#description')
            .simulate('change', { target: { value: product.description, name: 'description' } });
          // mock the router
         Router.router = { push: jest.fn() };
         wrapper.find('form').simulate('submit');
         await wait(50);
         expect(Router.router.push).toHaveBeenCalled();
         expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/product', query: { id: 'abc123' } });
    });
});