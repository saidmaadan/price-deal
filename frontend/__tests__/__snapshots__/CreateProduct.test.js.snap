// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1e7i5cp-0 jHGUAN"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter A Description"
        required={true}
        value=""
      />
    </label>
    <label
      htmlFor="originalprice"
    >
      Original Price
      <input
        id="originalPrice"
        name="originalPrice"
        onChange={[Function]}
        placeholder="originalPrice"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="discount"
    >
      Discount
      <input
        id="discount"
        name="discount"
        onChange={[Function]}
        placeholder="discount"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
